x-pack/plugins/apm/server/lib/helpers/get_document_sources.ts

```diff
@@ -57,25 +66,50 @@ export async function getDocumentSources({
   kuery,
   enableServiceTransactionMetrics,
   enableContinuousRollups,
+  environment,
+  serviceName,
 }: {
   apmEventClient: APMEventClient;
   start: number;
   end: number;
   kuery: string;
   enableServiceTransactionMetrics: boolean;
   enableContinuousRollups: boolean;
+  environment?: Environment;
+  serviceName?: string;
 }) {
   const currentRange = rangeQuery(start, end);
   const diff = end - start;
   const kql = kqlQuery(kuery);
   const beforeRange = rangeQuery(start - diff, end - diff);
+  const serviceFilter: QueryDslQueryContainer[] | undefined = serviceName
+    ? [
+        {
+          term: {
+            [SERVICE_NAME]: serviceName,
+          },
+        },
+        ...(environment &&
+        environment !== 'ENVIRONMENT_ALL' &&
+        environment !== 'ENVIRONMENT_NOT_DEFINED'
+          ? [
+              {
+                term: {
+                  [SERVICE_ENVIRONMENT]: environment,
+                },
+              },
+            ]
+          : []),
```

Comment:
I think you could use [getEnvironmentEsField](https://github.com/elastic/kibana/blob/main/x-pack/plugins/apm/common/environment_filter_values.ts#L55)

Reply:
Yeah, it builds the key-value pair needed for the terms agg. Thanks!